sdk/daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Upgrading.scala

```diff
@@ -375,31 +385,35 @@ abstract class TypecheckUpgradesUtils(
       arg: Upgrading[Map[K, V]],
       handler: (K, V) => UpgradeError.Error,
       filter: (K, V) => Boolean = (_: K, _: V) => true,
-  ): Try[(Map[K, Upgrading[V]], Map[K, V])] = {
+  ): UpgradeM[CheckDeleted[K, V]] = {
     val (deletedV, existingV, newV) = extractDelExistNew(arg)
     deletedV.view.filter((kv: (K, V)) => filter(kv._1, kv._2)).headOption match {
       case Some((k, v)) => fail(handler(k, v))
-      case _ => Success((existingV, newV))
+      case _ => succeed(CheckDeleted(existingV, newV))
     }
   }
 
   protected def checkLfVersions(
       arg: Upgrading[LanguageVersion]
-  ): Try[Unit] = {
+  ): UpgradeM[Unit] = {
     import Ordering.Implicits._
     if (arg.past > arg.present)
       fail(UpgradeError.DecreasingLfVersion(arg.past, arg.present))
     else
-      Success(())
+      succeed(())
   }
 
 }
 
 object TypecheckUpgrades {
+  type PartialEither[X] = Either[UpgradeError, X]
+  type UpgradeM[A] = WriterT[PartialEither, Chain[UpgradeError], A]
```

Comment:
So `Chain[UpgradeError]` are the warn ones I guess, I'd document this somewhere, as from the types it's harder to tell since we also represent with `UpgradeError` the failure mode of this writer